### ISSUE TITLE
Speed slower than pytorch conv2d with groups

### ISSUE THREAD

**Issue by ybzhou** [szagoruyko/pyinn#11]
I used this for replace pytorch's conv2d where I also set groups=input_feat, this implementation is taking almost twice as much time as in pytorch's.

**Comment by szagoruyko** [comment]
can you post your benchmark script

**Comment by ybzhou** [comment]
I was not benchmarking, I simply substitute nn.Conv2d with Conv2dDepthwise, my conv structure looks something like the following

`
layers = [
            nn.BatchNorm2d(in_feat),
            nn.Conv2d(in_feat,
                      in_feat,
                      kernel_size=kernel_size,
                      stride=stride,
                      padding=(f_h // 2, f_w // 2) if pad else 0,
                      groups=in_feat,
                      bias=False),

            nn.BatchNorm2d(in_feat),
            nn.Conv2d(in_feat,
                      out_feat,
                      kernel_size=(1,1),
                      stride=(1,1),
                      bias=False
                      ),

            nn.BatchNorm2d(out_feat),
            nn.ReLU()]
        self.sep_conv = nn.Sequential(*layers)
`
the first nn.conv2d is replaced by Conv2dDepthwise

**Comment by szagoruyko** [comment]
@ybzhou right, so can you provide a benchmark script with this network, and how you time it - remember you need to run pyinn once to compile, and time it after that. Compilation might take a few sec per kernel.

**Comment by szagoruyko** [comment]
here is an example without a network, pyinn is faster than cudnn6 pytorch convolution ~23 times:

```python
import torch
import torch.nn.functional as F
from pyinn import conv2d_depthwise
from torch.autograd import Variable
from torchnet.meter import TimeMeter

N = 1024

x = Variable(torch.randn(256, N, 14, 14).cuda())
w = Variable(torch.randn(N, 1, 3, 3).cuda())

y = F.conv2d(x, w, padding=1, groups=x.size(1))
y = conv2d_depthwise(x, w, padding=1)

meter = TimeMeter('s')

for i in range(10):
    y = F.conv2d(x, w, padding=1, groups=x.size(1))
    torch.cuda.synchronize()

t0 = meter.value()

meter = TimeMeter('s')

for i in range(10):
    y = conv2d_depthwise(x, w, padding=1)
    torch.cuda.synchronize()

t1 = meter.value()

print('torch time', t0)
print('pyinn time', t1)

print(t0 / t1)
```

output on Maxwell Titan X

```
('torch time', 1.0383970737457275)
('pyinn time', 0.04490780830383301)
23.1228624368
```

**Comment by szagoruyko** [comment]
@ybzhou see https://github.com/szagoruyko/pyinn/pull/12 for another example. feel free to reopen with repro script.

**Comment by ybzhou** [comment]
the forward is faster but the backward is a lot slower, following is the code for reproducing
I got the result on K80

```
import torch
import time
from torch import nn
from torch.autograd import Variable
from pyinn.modules import Conv2dDepthwise

in_feat = 32
out_feat = 64
batch_size = 256
stride = (2,1)
kernel_size = (7,3)

x = Variable(torch.randn(batch_size, in_feat, 40, 400)).cuda()
t = Variable(torch.randn(batch_size)).cuda()

print(kernel_size, stride)
layers = [nn.BatchNorm2d(in_feat),
          nn.Conv2d(in_feat,
                    in_feat,
                    kernel_size=kernel_size,
                    stride=stride,
                    padding=0,
                    groups=in_feat,
                    bias=False),

          nn.BatchNorm2d(in_feat),
          nn.Conv2d(in_feat,
                    out_feat,
                    kernel_size=(1,1),
                    stride=(1,1),
                    bias=False
                    ),

          nn.BatchNorm2d(out_feat),
          nn.ReLU()]

torch_conv = nn.Sequential(*layers)
torch_conv = torch_conv.cuda()

layers = [nn.BatchNorm2d(in_feat),
          Conv2dDepthwise(in_feat,
                          kernel_size=kernel_size,
                          stride=stride,
                          padding=0,
                          bias=False),

          nn.BatchNorm2d(in_feat),
          nn.Conv2d(in_feat,
                    out_feat,
                    kernel_size=(1,1),
                    stride=(1,1),
                    bias=False
                    ),

          nn.BatchNorm2d(out_feat),
          nn.ReLU()]

pyinn_conv = nn.Sequential(*layers)
pyinn_conv = pyinn_conv.cuda()

print('forward1')
t_start = time.time()
for i in range(10):
    y1 = torch_conv(x)
    torch.cuda.synchronize()

pytorch_forward_t = time.time()-t_start

y1 = y1.sum(1).sum(1).sum(1)
loss1 = nn.MSELoss()(y1, t)
print('backward1')
t_start = time.time()
loss1.backward()
pytorch_backward_t = time.time() - t_start

print('forward2')
t_start = time.time()
for i in range(10):
    y2 = pyinn_conv(x)
    torch.cuda.synchronize()
pyinn_forward_t = time.time()-t_start

y2 = y2.sum(1).sum(1).sum(1)
loss2 = nn.MSELoss()(y2, t)
print('backward2')
t_start = time.time()
loss2.backward()
pyinn_backward_t = time.time() - t_start

print('batch size: {}, kernel size: {}, stride size: {}, '
      'pytorch forward time: {:.4f}s '
      'pyinn forward time: {:.4f}s '
      'pytorch backward time: {:.4f}s '
      'pyinn backward time: {:.4f}s'.format(
    batch_size, kernel_size, stride,
    pytorch_forward_t, pyinn_forward_t,
    pytorch_backward_t, pyinn_backward_t
))
print(pytorch_forward_t/pyinn_forward_t,
      pytorch_backward_t/pyinn_backward_t)

```

```
batch size: 256, kernel size: (7, 3), stride size: (2, 1), pytorch forward time: 2.4811s pyinn forward time: 1.5070s pytorch backward time: 0.0037s pyinn backward time: 0.1597s
1.646389619890211 0.022934876056741653
```

**Comment by szagoruyko** [comment]
@ybzhou you are not benchmarking correctly:
 * CUDA calls are asynchronous, you need to call `torch.cuda.synchronize` before every call
 * pyinn needs time to compile kernels in forward and backward.

with correct benchmarks you are still right though, backward w.r.t. weights is very slow because of the way it's written
